sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantNullCheck.cs

```diff
@@ -0,0 +1,106 @@
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.CSharp;
+using SonarAnalyzer.ShimLayer.CSharp;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class RedundantNullCheck : RedundantNullCheckBase<BinaryExpressionSyntax>
+    {
+        private const string MessageFormat = "Remove this unnecessary null check; 'is' returns false for nulls.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                CheckAndExpression,
+                SyntaxKind.LogicalAndExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                CheckOrExpression,
+                SyntaxKind.LogicalOrExpression);
+        }
+
+        protected override SyntaxNode GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left.RemoveParentheses();
+
+        protected override SyntaxNode GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right.RemoveParentheses();
+
+        private SyntaxNode GetNullCheckVariableForKind(SyntaxNode node, SyntaxKind kind) {
+            if (RemoveParentheses(node) is BinaryExpressionSyntax binaryExpression && binaryExpression.IsKind(kind)) {
+                var leftNode = GetLeftNode(binaryExpression);
+                var rightNode = GetRightNode(binaryExpression);
+
+                if (leftNode.IsKind(SyntaxKind.NullLiteralExpression)) {
+                    return rightNode;
+                }
+                if (rightNode.IsKind(SyntaxKind.NullLiteralExpression))
+                {
+                    return leftNode;
+                }
+            }
+            return null;
+        }
+
+        protected override SyntaxNode GetNullCheckVariable(SyntaxNode node) => GetNullCheckVariableForKind(node, SyntaxKind.EqualsExpression);
+
+        protected override SyntaxNode GetNonNullCheckVariable(SyntaxNode node) => GetNullCheckVariableForKind(node, SyntaxKind.NotEqualsExpression);
+
+        protected override SyntaxNode GetIsOperatorCheckVariable(SyntaxNode node)
+        {
+            var innerExpression = RemoveParentheses(node);
+            if (innerExpression is BinaryExpressionSyntax binaryExpression && binaryExpression.IsKind(SyntaxKind.IsExpression))
+            {
+                return GetLeftNode(binaryExpression);
+            } else if (innerExpression.IsKind(SyntaxKindEx.IsPatternExpression))
```

Comment:
Put `else if` on new line 👍 

However, the expression cannot be replaced by `innerExpression is IsPatternExpressionSyntaxWrapper isPatternExpression` because it is part of the ShimLayer, and `IsPatternExpressionSyntaxWrapper` is not a subclass of `SyntaxNode`, but a wrapper on it